core/src/main/java/org/apache/iceberg/FastAppend.java

```diff
@@ -178,8 +184,25 @@ public Object updateEvent() {
 
   @Override
   protected void cleanUncommitted(Set<ManifestFile> committed) {
-    if (newManifest != null && !committed.contains(newManifest)) {
-      deleteFile(newManifest.path());
+    if (newManifests != null) {
+      // Delete those manifests are not contained in the committed set and keep others as here.
+      // This is needed for manifests cleanup especially in transaction mode, for example:
+      //   Transaction txn = beginTransaction(...)
+      //   // success operation and call this cleanUncommittedAppends
+      //   txn.newFastAppend().appendFile(...).commit();
+      //   some other operations ...
+      //   // commit failed need to clean up those successes committed manifests
```

Comment:
'Commit fails and needs to clean up newManifests' ?  (they are not committed , right?)

Reply:
Updated